### PR TITLE
Support SANs in TLS cert to define user permissions

### LINKED ISSUE #964: Permit mapping user permissions from the client certificate SAN
- [ ] Defect
 - [x] Feature Request or Change Proposal

#### Use Case:

For server-to-server communication, using the email claim or DN syntax may not be the most appropriate for Identifying a user.  In my use case and others, the proper identifier of the client is present in the subject alternative name (SAN) field in the client cert.

It would be great to allow SANs to be mapped to users when the `verify_and_map` TLS option is enabled.
 
#### Proposed Change:

Permit specifying the mapped user with values pulled from the client certificate SAN, as was done in this commit: https://github.com/wallyqs/gnatsd/commit/502adb3ed9381aa77de0951894b1ab65c9c555d2

Moreover, it would be more robust to not always select the first value from (email ∪ SAN ∪ DN) and use that as the user name, failing if that doesn't exist.  Instead, each candidate username (email, each SAN, and the DN) can be checked for existence, and the first one to exist would be returned, or fail only if all options do not exist.

This change would also be backwards compatible with the current implementation.

#### Who Benefits From The Change(s)?

Any users who wishe to authenticate and map users authorization via mTLS certs, but who may not have control over how those certs are issued (for me, we have a central application authority which issues certificates, and changing their format is out of my control).
 
#### Alternative Approaches

We could add additional TLS configs to specify exactly what to pull from a certificate to determine a user.  However, this would require a more complex TLS configuration and isn't as flexible.

### PR DESCRIPTION
**Pull request by wallyqs**
- [ ] Documentation added (if applicable)
 - [X] Tests added
 - [X] Branch rebased on top of current master (`git pull --rebase origin master`)
 - [x] Changes squashed to a single commit
 - [X] Build is green in Travis CI
 - [X] You have certified that the contribution is your original work and that you license the work to the project under the [Apache 2 license](https://github.com/nats-io/gnatsd/blob/master/LICENSE)

Fixes #964 

### Changes proposed in this pull request:

 - Support SANs found in a cert for auth
 - Allow multiple emails and/or SANs to be defined, using the first one that is valid
 - Use subject cert when neither emails or SANs were valid
 - Bug fix to not logging properly user on failed auth attempt

/cc @nats-io/core

### CODE REVIEW

File: test/tls_test.go [hunk 1]
```diff
@@ -248,6 +257,178 @@ func TestTLSClientCertificateSANsBasedAuth(t *testing.T) {
 	defer devc.Close()
 
 	// prod: Using SAN 'app.nats.prod' with all permissions.
+	//
+	//    Subject: CN = www.nats.io
+	//
+	//    X509v3 Subject Alternative Name:
+	//        DNS:app.nats.prod, DNS:*.app.nats.prod
```
Comment (derekcollison) [inline test/tls_test.go]: How do we resolve multiple entries?

Comment (wallyqs) [reply test/tls_test.go]: The cert defines the order in which multipe entries will be tried.  For example, in this case the cert has two SANs defined which will result in the following array always in this order:

```ruby
DNSNames = ["app.nats.prod", "*.app.nats.prod"]
```

We then try each until we find a SANs that is present in the config.  With the config below for example, `app.nats.prod` would be looked first but would fail since not present in the config and the SAN `*.app.nats.prod` would be used instead.  If neither of the SANs are present then in the config then the subject in the cert using the DN syntax would be used (`CN=www.nats.io` in this case).

```hcl
users [
  { user = "admin@nats.io" }    # Email is tried first, if no emails then other methods are tried
  { user = "*.app.nats.prod" }  # Use one of SANs if present
  { user = "CN=www.nats.io" }   # Defaul to use the subject in the cert if no other method available
]
```

Comment (derekcollison) [reply test/tls_test.go]: ok I think I follow.

Comment (twrobel3) [reply test/tls_test.go]: Yup, that's exactly the type of functionality I was looking for.  Thanks again!

File: server/auth.go [hunk 2]
```diff
@@ -320,7 +320,7 @@ func (s *Server) isClientAuthorized(c *client) bool {
 				var ok bool
 				user, ok = s.users[u]
 				if !ok {
-					c.Debugf("User in cert [%q], not found", euser)
+					c.Debugf("User in cert [%q], not found", u)
```
Comment (ColinSullivan1) [inline server/auth.go]: I'd rephrase this a bit, User [%q] not found in cert.

**Comment by ColinSullivan1** [comment]
A little late, but LGTM. ;)
